### PR TITLE
ceph: prevent voluntary mon drain while another mon is failing over

### PR DESCRIPTION
**Pull request by sp98**
With simultaneous node drains, a mon can go down while another mon is failing over. This results in two mons down and ceph become inaccessible.
This PR updates mon PDB to use maxUnavailable=1. Its updates the mon pdb to use maxUavailable=0 while a mon is failing over.

Signed-off-by: Santosh Pillai <sapillai@redhat.com>

**Description of your changes:**

**Which issue is resolved by this Pull Request:**
Resolves #

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/master/development-flow.html#commit-structure).
- [ ] **Skip Tests for Docs**: Add the flag for skipping the build if this is only a documentation change. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for the flag.
- [ ] **Skip Unrelated Tests**: Add a flag to run tests for a specific storage provider. See [test options](https://github.com/rook/rook/blob/master/INSTALL.md#test-storage-provider).
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/master/development-flow.html#submitting-a-pull-request)
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.

### CODE REVIEW

File: pkg/operator/ceph/cluster/mon/drain.go [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+	_, err = ctx.Clientset.PolicyV1beta1().PodDisruptionBudgets(namespace).Update(context.TODO(), monPDB, metav1.UpdateOptions{})
+	if err != nil {
+		return errors.Wrapf(err, "failed to update MaxUnavailable for mon PDB %q", monPDB.Name)
+	}
+
+	return nil
+}
+
+func getMonPDB(ctx *clusterd.Context, namespace string) (*policyv1beta1.PodDisruptionBudget, error) {
+	existingPDB, err := ctx.Clientset.PolicyV1beta1().PodDisruptionBudgets(namespace).Get(context.TODO(), "rook-ceph-mon-pdb", metav1.GetOptions{})
```
Comment (leseb) [inline pkg/operator/ceph/cluster/mon/drain.go:59]:
```suggestion
	existingPDB, err := ctx.Clientset.PolicyV1beta1().PodDisruptionBudgets(namespace).Get(context.TODO(), monPDBName, metav1.GetOptions{})
```

File: pkg/operator/ceph/cluster/mon/health.go [hunk 2]
```diff
@@ -237,10 +237,26 @@ func (c *Cluster) failMon(monCount, desiredMonCount int, name string) {
...
+			}
+		}
+
 		// bring up a new mon to replace the unhealthy mon
 		if err := c.failoverMon(name); err != nil {
 			logger.Errorf("failed to failover mon %q. %v", name, err)
 		}
+
+		if c.spec.DisruptionManagement.ManagePodBudgets {
+			logger.Info("allow voluntary mon drian after fail over")
```
Comment (leseb) [inline pkg/operator/ceph/cluster/mon/health.go:254]:
```suggestion
			logger.Info("allow voluntary mon drain after failover")
```

File: pkg/operator/ceph/cluster/mon/health.go [hunk 3]
```diff
@@ -237,10 +237,26 @@ func (c *Cluster) failMon(monCount, desiredMonCount int, name string) {
...
+				logger.Errorf("failed to block mon drain. %v", err)
+			}
+		}
+
 		// bring up a new mon to replace the unhealthy mon
 		if err := c.failoverMon(name); err != nil {
 			logger.Errorf("failed to failover mon %q. %v", name, err)
 		}
+
+		if c.spec.DisruptionManagement.ManagePodBudgets {
```
Comment (leseb) [inline pkg/operator/ceph/cluster/mon/health.go:253]: Just because we only log errors, I believe the commit should explain that we are in best-effort mode.

**Review by leseb — CHANGES_REQUESTED** [review]

File: pkg/operator/ceph/cluster/mon/health.go (hunk 3 above)
Comment (sp98) [reply pkg/operator/ceph/cluster/mon/health.go:253]: updated commit message to indicate that we are using best effort.

File: pkg/operator/ceph/cluster/mon/drain_test.go [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+func TestBlockMonDrainFailure(t *testing.T) {
+	clientset := test.New(t, 1)
+	ctx := &clusterd.Context{
+		Clientset: clientset,
+	}
+	// no mon PDB created
+	err := blockMonDrain(ctx, mockPDBNamespace)
+	assert.Error(t, err)
+}
```
Comment (BlaineEXE) [inline pkg/operator/ceph/cluster/mon/drain_test.go:102]: Instead of creating a bunch of separate tests, why not create a `Test_blockMonDrain` test and then use `t.Run` to run subtests?

Ditto for `Test_allowMonDrain`

Comment (sp98) [reply pkg/operator/ceph/cluster/mon/drain_test.go:102]: I feel there aren't enough cases to handle. There is one positive case that pdb was updated successfully and negative case that pdb update was failed because pdb was not found.  Adding them to a single test case with `t.Run` would make it a bit complicated. We would also need to delete the mon PDB to test the negative test case.

Comment (sp98) [reply pkg/operator/ceph/cluster/mon/drain_test.go:102]: I can have like below. Looks like an over kill. What do you think?

```
func Test_AllowMonDrain(t *testing.T) {
	clientset := test.New(t, 1)
	ctx := &clusterd.Context{
		Clientset: clientset,
	}
	tests := []struct {
		name         string
		createMonPDB bool
		wantErr      bool
	}{
		{"failed to allow mon drain", false, true},
		{"allow mon drain", true, false},
	}

	for _, tt := range tests {
		if tt.createMonPDB {
			err := createMonPDB(t, ctx)
			assert.NoError(t, err)
		}
		t.Run(tt.name, func(t *testing.T) {
			// change MaxUnavailable mon PDB to 1
			err := allowMonDrain(ctx, mockPDBNamespace)
			if tt.wantErr {
				assert.Error(t, err)
				return
			}
			assert.NoError(t, err)
			existingPDB, err := ctx.Clientset.PolicyV1beta1().PodDisruptionBudgets(mockPDBNamespace).Get(context.TODO(), monPDBName, metav1.GetOptions{})
			assert.NoError(t, err)
			assert.Equal(t, 1, int(existingPDB.Spec.MaxUnavailable.IntValue()))
		})
	}

}```

Comment (sp98) [reply pkg/operator/ceph/cluster/mon/drain_test.go:102]: updated the tests. Please take a look.

File: pkg/operator/ceph/cluster/mon/drain_test.go [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+			},
+			MaxUnavailable: &intstr.IntOrString{IntVal: 1},
+		},
+	}
+
+	_, err := ctx.Clientset.PolicyV1beta1().PodDisruptionBudgets(mockPDBNamespace).Create(context.TODO(), pdbObj, metav1.CreateOptions{})
+
+	return err
+}
+func Test_AllowMonDrain(t *testing.T) {
```
Comment (leseb) [inline pkg/operator/ceph/cluster/mon/drain_test.go:54]:
```suggestion
func TestAllowMonDrain(t *testing.T) {
```

File: pkg/operator/ceph/cluster/mon/drain_test.go [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+			assert.NoError(t, err)
+			existingPDB, err := ctx.Clientset.PolicyV1beta1().PodDisruptionBudgets(mockPDBNamespace).Get(context.TODO(), monPDBName, metav1.GetOptions{})
+			assert.NoError(t, err)
+			assert.Equal(t, 1, int(existingPDB.Spec.MaxUnavailable.IntValue()))
+		})
+	}
+
+}
+
+func Test_BlockMonDrain(t *testing.T) {
```
Comment (leseb) [inline pkg/operator/ceph/cluster/mon/drain_test.go:89]:
```suggestion
func TestBlockMonDrain(t *testing.T) {
```

File: pkg/operator/ceph/disruption/clusterdisruption/reconcile.go [hunk 7]
```diff
@@ -267,18 +277,33 @@ func (c *ClusterMap) GetClusterNamespaces() []string {
...
-		client.MatchingLabels{k8sutil.AppAttr: "rook-ceph-osd-pdb"})
+		client.MatchingLabels{k8sutil.AppAttr: legacyOSDPDBLabel})
 	if err != nil && !kerrors.IsNotFound(err) {
-		return errors.Wrap(err, "failed to delete all the legacy blocking podDisruptionBugets")
+		return errors.Wrapf(err, "failed to delete legacy OSD PDBs with label %q", legacyOSDPDBLabel)
 	}
 	return nil
 }
+
+func (r *ReconcileClusterDisruption) deleteLegacyPDBForMons(namespace string) error {
```
Comment (leseb) [inline pkg/operator/ceph/disruption/clusterdisruption/reconcile.go:296]: What makes the legacy mon PDB actually "legacy"? What the difference between this one and the new one (if any)?

**Review by leseb — CHANGES_REQUESTED** [review]

File: pkg/operator/ceph/disruption/clusterdisruption/reconcile.go (hunk 7 above)
Comment (sp98) [reply pkg/operator/ceph/disruption/clusterdisruption/reconcile.go:296]: The current mon PDB uses  `minAvailable= int32(float64((monCount + 1) / 2))`.  So for 3 mons, minAvaliable=2, for 5 mons minAvaliable=2 and likewise.  For this PR, we decided to play it safer and change it to maxUnavailable=1 (irrespective of the number of mons). 

In case of cluster upgrade, the mon pdb  with `minAvailable` should be the legacy pdb. We need to delete that pdb.

**Review by leseb — APPROVED** [review]

File: pkg/operator/ceph/disruption/clusterdisruption/reconcile.go (hunk 7 above)
Comment (nehaberry) [reply pkg/operator/ceph/disruption/clusterdisruption/reconcile.go:296]: @sp98 Isn't the legacy blocking PDB only for OSD pods? Are you going to change the PDB design then ?

$ oc get pdb -n openshift-storage
NAME                                              MIN AVAILABLE   MAX UNAVAILABLE   ALLOWED DISRUPTIONS   AGE
rook-ceph-mds-ocs-storagecluster-cephfilesystem   1               N/A               1                     30h
rook-ceph-mon-pdb                                 2               N/A               1                     30h
rook-ceph-osd                                     N/A             1                 1                     43m
[nberry@localhost logs]$

File: pkg/operator/ceph/cluster/mon/drain_test.go (hunk 4 above)
Comment (BlaineEXE) [reply pkg/operator/ceph/cluster/mon/drain_test.go]: It's no difference between creating two tests versus writing sub-tests with `t.Run`. Using sub-tests is a little more clear, however. I think you confuse my request to use `t.Run` as a request to use the `for` loop structure you proposed, but you can use `t.Run` at any point in a test. See here for example https://github.com/rook/rook/blob/736f800c1d47d3d37a2929d32a654d15836ba478/pkg/operator/ceph/cluster/osd/update_test.go#L184

File: pkg/operator/ceph/cluster/mon/drain_test.go [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+			if tt.wantErr {
+				assert.Error(t, err)
+				return
+			}
+			assert.NoError(t, err)
+			existingPDB, err := ctx.Clientset.PolicyV1beta1().PodDisruptionBudgets(mockPDBNamespace).Get(context.TODO(), monPDBName, metav1.GetOptions{})
+			assert.NoError(t, err)
+			assert.Equal(t, 0, int(existingPDB.Spec.MaxUnavailable.IntValue()))
+		})
+	}
```
Comment (BlaineEXE) [inline pkg/operator/ceph/cluster/mon/drain_test.go:119]: I think this is a lot more confusing to read with the loop. Why not create two separate calls to `t.Run` instead and do the setup for each as needed. I posted another comment with a link where you can see what I mean.

**Review by BlaineEXE — CHANGES_REQUESTED** [review]

File: pkg/operator/ceph/cluster/mon/drain_test.go (hunk 8 above)
Comment (BlaineEXE) [reply pkg/operator/ceph/cluster/mon/drain_test.go:119]: ^ https://github.com/rook/rook/pull/7442#discussion_r601608664

Comment (sp98) [reply pkg/operator/ceph/cluster/mon/drain_test.go]: Thanks for sharing the link. I've updated the tests.

File: pkg/operator/ceph/disruption/clusterdisruption/reconcile.go (hunk 7 above)
Comment (sp98) [reply pkg/operator/ceph/disruption/clusterdisruption/reconcile.go:296]: we can discuss offline or on a call. @nehaberry

File: pkg/operator/ceph/disruption/clusterdisruption/mon.go [hunk 9]
```diff
@@ -31,14 +31,8 @@ const (
 	pdbName = "rook-ceph-mon-pdb"
 )
 
-// setting minAvailable for mons at floor((n+1)/2) an logging and error if n is even
-// monCount - minAvailable
-// 1,2      - 0 (no HA)
-// even     - n -1
-// odd      - floor((n+1)/2)
 func (r *ReconcileClusterDisruption) reconcileMonPDB(cephCluster *cephv1.CephCluster) error {
```
Comment (travisn) [inline pkg/operator/ceph/disruption/clusterdisruption/mon.go:34]: Does the disruption controller still need to reconcile the mon pdb? Seems like we just want the cluster reconciler to manage the mon pdb so we can avoid any conflicts between the controllers. For example, if a mon failover is in progress, we don't want the disruption controller to reconcile the pdb.

File: pkg/operator/ceph/disruption/clusterdisruption/reconcile.go (hunk 7 above)
Comment (travisn) [reply pkg/operator/ceph/disruption/clusterdisruption/reconcile.go:296]: I don't think we actually need to worry about a legacy pdb. The pattern I'm thinking is that when the cluster controller reconciles the mons, it will always update the mon pdb with the expected pdb. So the first time the operator runs with this change, the pdb would automatically be updated, and no need to delete the previous pdb.

**Review by travisn — CHANGES_REQUESTED** [review]

File: pkg/operator/ceph/disruption/clusterdisruption/mon.go (hunk 9 above)
Comment (sp98) [reply pkg/operator/ceph/disruption/clusterdisruption/mon.go:34]: since we changed from dynamic `minAvailable` to a static `maxUnavailable=1`, the reconciler won't change the `maxUnavailable` after creating it initially. See (https://github.com/rook/rook/blob/c2a62b63f9414396a1ba1793210c85eb871baf24/pkg/operator/ceph/disruption/clusterdisruption/static_pdb.go#L37-L48).  So disruption controller reconciler won't interfere and change the mon pdbs while mons are failing over.

File: pkg/operator/ceph/disruption/clusterdisruption/reconcile.go (hunk 7 above)
Comment (sp98) [reply pkg/operator/ceph/disruption/clusterdisruption/reconcile.go:296]: We can discuss more on this.  I would suggest having the `cluster disruption controller` to continue reconciling the mon pdbs, rather than moving it to `cluster controller`.

Comment (travisn) [reply pkg/operator/ceph/disruption/clusterdisruption/reconcile.go:296]: Per our discussion, let's move the mon pdb reconcile code into the mon reconcile, and no need for it in the disruption reconciler. For example, around [line 300 of mon.go](https://github.com/rook/rook/blob/4d06ad1155e8eacdb44acb97cb46c88d98555716/pkg/operator/ceph/cluster/mon/mon.go#L300) after the mons are reconciled, we can insert a method that reconciles the mon pdb. The pdb would follow the pattern to create the pdb if it doesn't exist, and update it if it already exists (even if the spec didn't change). 

The `updateMonPDB()` method would just need to follow the pattern to generate the mon pdb spec and create the pdb if it doesn't exist yet, so some implementation will move over from the disruption controller.

File: pkg/operator/ceph/cluster/mon/drain.go [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+
+	"github.com/pkg/errors"
+	"github.com/rook/rook/pkg/clusterd"
+	policyv1beta1 "k8s.io/api/policy/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+)
+
+const (
+	monPDBName = "rook-ceph-mon-pdb"
```
Comment (BlaineEXE) [inline pkg/operator/ceph/cluster/mon/drain.go:30]: Nit: I would call this `rook-ceph-mons`. The type is already a PDB, so it becomes redundant information to put `pdb` in the name of a PDB resource.

@travisn @leseb do you have other thoughts?

Comment (leseb) [reply pkg/operator/ceph/cluster/mon/drain.go:30]: If this is not displayed to the user then I'm fine with `rook-ceph-mons`.

Comment (travisn) [reply pkg/operator/ceph/cluster/mon/drain.go:30]: Agreed that name would be better, but I would suggest we wait for that in a separate PR. Let's keep this PR simpler to get it in sooner. If we rename it, then we do have to worry about deleting a legacy pdb.

Comment (sp98) [reply pkg/operator/ceph/cluster/mon/drain.go:32]: Yes, if we change then we need to delete the old mon PDBs as well. Keeping it `rook-ceph-mon-pdb` for now.

File: pkg/operator/ceph/disruption/clusterdisruption/reconcile.go (hunk 7 above)
Comment (sp98) [reply pkg/operator/ceph/disruption/clusterdisruption/reconcile.go]: moved mon pdb reconcile code into mon reconcile. Testing looks good. This is ready for review now.

File: pkg/operator/ceph/cluster/mon/drain.go [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+
+func (c *Cluster) createOrUpdateMonPDB() (controllerutil.OperationResult, error) {
+	pdb := &policyv1beta1.PodDisruptionBudget{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      monPDBName,
+			Namespace: c.Namespace,
+		}}
+
+	mutateFunc := func() error {
+		pdb.Spec = policyv1beta1.PodDisruptionBudgetSpec{
```
Comment (travisn) [inline pkg/operator/ceph/cluster/mon/drain.go:57]: Can we just set the spec on the struct above? Seems like we don't need a mutate func to change anything.

File: pkg/operator/ceph/cluster/mon/drain.go [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+	err := c.context.Client.Update(context.TODO(), pdb)
+	if err != nil {
+		return errors.Wrapf(err, "failed to update pdb %q", pdb.Name)
+	}
+	return nil
+}
+
+func (c *Cluster) getMonPDB(request types.NamespacedName) (*policyv1beta1.PodDisruptionBudget, error) {
+	existingPDB := &policyv1beta1.PodDisruptionBudget{}
+	err := c.context.Client.Get(context.TODO(), request, existingPDB)
```
Comment (travisn) [inline pkg/operator/ceph/cluster/mon/drain.go:98]: I don't think we need to get the pdb, we just need to call `createOrUpdateMonPDB()` with the desired pdb spec

File: pkg/operator/ceph/cluster/mon/drain.go [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+	}
+
+	op, err := c.createOrUpdateMonPDB()
+	if err != nil {
+		return errors.Wrapf(err, "failed to reconcile mon pdb on op %q", op)
+	}
+	return nil
+}
+
+func (c *Cluster) createOrUpdateMonPDB() (controllerutil.OperationResult, error) {
```
Comment (travisn) [inline pkg/operator/ceph/cluster/mon/drain.go:49]: What if this method took the maxUnavailable as a param, then everyone could just call this?
```suggestion
func (c *Cluster) createOrUpdateMonPDB(maxUnavailable int) (controllerutil.OperationResult, error) {
```

File: pkg/operator/ceph/cluster/mon/health.go [hunk 14]
```diff
@@ -237,10 +238,26 @@ func (c *Cluster) failMon(monCount, desiredMonCount int, name string) {
 			logger.Errorf("failed to remove mon %q. %v", name, err)
 		}
 	} else {
+		if c.spec.DisruptionManagement.ManagePodBudgets {
```
Comment (travisn) [inline pkg/operator/ceph/cluster/mon/health.go:241]: Instead of checking for ManagePodBudgets here, I'd suggest passing the setting to the methods on the pdb handler so they can enable or disable the pdb. It will make the code a bit simpler for the caller here.

File: pkg/operator/ceph/cluster/mon/drain.go [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+	monPDBName = "rook-ceph-mon-pdb"
+)
+
+func (c *Cluster) reconcileMonPDB() error {
+	monCount := c.spec.Mon.Count
+	if monCount <= 2 {
+		logger.Debug("managePodBudgets is set, but mon-count <= 2. Not creating a disruptionbudget for Mons")
+		return nil
+	}
+
```
Comment (travisn) [inline pkg/operator/ceph/cluster/mon/drain.go:41]: If the pdb is not enabled, we also need to delete the pdb as well.

**Review by travisn — CHANGES_REQUESTED** [review]

File: pkg/operator/ceph/cluster/mon/drain.go (hunk 15 above)
Comment (sp98) [reply pkg/operator/ceph/cluster/mon/drain.go:41]: How about handling deletion of all PDB (mon and OSD) in a separate PR?  (if we handle mon here, we would need to handle osds as well else `ManagePodBudgets=false` would only delete mon PDB)

File: pkg/operator/ceph/cluster/mon/drain.go (hunk 11 above)
Comment (sp98) [reply pkg/operator/ceph/cluster/mon/drain.go:62]: `controllerutil.CreateOrUpdate` needs `mutate` func to check resource `spec` has changed or not. It updates the resource if the spec has changed. `blockMonDrain` and `allowMonDrain` are also using this now. So we need `mutate` func here.

File: pkg/operator/ceph/cluster/mon/drain.go (hunk 12 above)
Comment (sp98) [reply pkg/operator/ceph/cluster/mon/drain.go]: updated to use `createOrUpdateMonPDB`.

File: pkg/operator/ceph/cluster/mon/drain.go (hunk 13 above)
Comment (sp98) [reply pkg/operator/ceph/cluster/mon/drain.go]: done.

File: pkg/operator/ceph/cluster/mon/health.go (hunk 14 above)
Comment (sp98) [reply pkg/operator/ceph/cluster/mon/health.go]: checking whether `ManagePodBudgets` is set inside the `allowMonDrain` and `blockMonDrain` functions.

File: pkg/operator/ceph/cluster/mon/drain.go (hunk 15 above)
Comment (travisn) [reply pkg/operator/ceph/cluster/mon/drain.go:46]: It's fine if that one setting is applied in different controllers for the two different PDBs, I don't see why the same controller has to delete the PDBs. This controller owns the mon pdb and the osd pdb is owned by the disruption controller. In a separate PR is fine though, it's not critical for this PR.

**Review by travisn — APPROVED** [review]

**Review by BlaineEXE — APPROVED** [review]
